algorithms/nautilus/grammartec/rule.cpp

```diff
@@ -0,0 +1,409 @@
+ * @param (ctx) Context
+ * @param (nonterm) String representation of nonterminal symbol
+ * @param (format) Format string (expression) of the rule for @p nonterm
+ *
+ * @details Nonterminal symbol must follow `{ABC:xyz}` format
+ *          where `ABC` indicates the nonterminal and `:xyz` is
+ *          an optional comment.
+ *          The format string is what the nonterminal should be
+ *          replaced with.
+ *          You can use any nonterminal symbols in this format.
+ */
+Rule::Rule(Context& ctx, const std::string& nonterm, const std::string& format) {
+  std::vector<RuleChild> children = Rule::Tokenize(format, ctx);
+  std::vector<NTermID> nonterms;
+
+  // Filter only NTerm items from children
+  for (RuleChild child: children) {
+    if (std::holds_alternative<NTerm>(child.value())) {
+      nonterms.emplace_back(std::get<NTerm>(child.value()));
+    }
+  }
+
+  _rule = PlainRule{ctx.AquireNTID(nonterm), children, nonterms};
+}
+
+/**
+ * @fn
+ * Describe this rule into a human-readable string.
+ * @brief Describe this rule
+ * @param (ctx) Context
+ * @return Human-readable string of this rule
+ */
+std::string Rule::DebugShow(Context& ctx) const {
+  return std::visit([&ctx](auto r) { return r.DebugShow(ctx); }, _rule);
+}
+
+/**
+ * @fn
+ * Remove every backslash before `{` and `}`
+ * @brief Unescape "\{" and "\}" in a string
+ * @param (bytes) String to unescape
+ * @return Unescaped string
+ */
+std::string Rule::Unescape(const std::string& bytes) {
+  if (bytes.size() < 2) {
+    // Nothing to escape
+    return bytes;
+  }
+
+  /* Convert "\{" into "{" and "\}" into "}" */
+  std::string res;
+  size_t i;
+  for (i = 0; i < bytes.size(); i++) {
+    if (bytes[i] == '\\' && bytes[i+1] == '{') {
+      res += "{";
+      i++;
+    } else if (bytes[i] == '\\' && bytes[i+1] == '}') {
+      res += "}";
+      i++;
+    } else {
+      res += bytes[i];
+    }
+  }
+
+  if (i < bytes.size()) {
+    res += bytes[bytes.size() - 1];
+  }
+
+  return res;
+}
+
+/**
+ * @fn
+ * Parse an expression of a rule and return a list of RuleChild.
+ * @brief Tokenize format string
+ * @param (format) String to tokenize
+ * @param (ctx) Context
+ * @throw exceptions::unreachable @p format is an invalid expression
+ * @return Children rules
+ */
+std::vector<RuleChild> Rule::Tokenize(const std::string& format, Context& ctx) {
+  static std::regex TOKENIZER(R"((\{[^}\\]+\})|((?:[^{\\]|\\\{|\\\}|\\)+))");
+
+  std::vector<RuleChild> r;
+  for (std::sregex_iterator it(format.begin(), format.end(), TOKENIZER), end;
+       it != end;
+       ++it) {
+    std::smatch m = *it;
+
+    if (m[1].matched) {
+      // NT: "{A:a}"
+      r.emplace_back(m.str(), ctx);
+    } else if (m[2].matched) {
+      // "abc\{def\}ghi" --> "abc{def}ghi"
+      r.emplace_back(Rule::Unescape(m.str()));
+    } else {
+      throw exceptions::unreachable(
+        "Unexpected capturing group", __FILE__, __LINE__
+      );
+    }
+  }
+
+  ctx = ctx;
```

Comment:
What's this line?

Reply:
Removed in [82c195c](https://github.com/fuzzuf/fuzzuf/pull/23/commits/82c195c436c8cfb04c86daa5abc249ced0e6b126)